### PR TITLE
KEYCLOAK-19374 Create entity implementations via annotation processor

### PR DESCRIPTION
**Pull request by hmlnarik**
Use of boxed types as started in 009d4ca4457 is finalized here
to enable storing data in a map. MapClientEntity methods are
reordered for the sake of grouping the collection-based
properties together and understanding the connections between those.

### CODE REVIEW

File: model/build-processor/src/main/java/org/keycloak/models/map/processor/GenerateEntityImplementationsProcessor.java [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+                super.println(x == null ? x : x.replaceAll("java.lang.", ""));
+            }
+        }) {
+            if (packageName != null) {
+                pw.println("package " + packageName + ";");
+            }
+
+            pw.println("import java.util.EnumMap;");
+            pw.println("import java.util.Objects;");
+            pw.println("public class " + mapSimpleClassName + (an.inherits().isEmpty() ? "" : " extends " + an.inherits()) + " implements " + className + " {");
```
Comment (mhajas) [inline model/build-processor/src/main/java/org/keycloak/models/map/processor/GenerateEntityImplementationsProcessor.java:162]: Should we add some warning that the class is generated and should not be changed manually?

File: model/map/src/main/java/org/keycloak/models/map/common/Serialization.java [hunk 2]
```diff
@@ -51,6 +52,7 @@
       .setSerializationInclusion(JsonInclude.Include.NON_NULL)
       .setVisibility(PropertyAccessor.ALL, Visibility.NONE)
       .setVisibility(PropertyAccessor.FIELD, Visibility.ANY)
+      .activateDefaultTyping(new LaissezFaireSubTypeValidator() /* TODO - see javadoc */, ObjectMapper.DefaultTyping.JAVA_LANG_OBJECT, JsonTypeInfo.As.PROPERTY)
```
Comment (mhajas) [inline model/map/src/main/java/org/keycloak/models/map/common/Serialization.java:55]: Why is this change necessary?

**Review by mhajas — COMMENTED** [review]
Thank you @hmlnarik! This looks great, maybe it will require some changes for more complicated  entities like Realm/User, but this can be added later, if necessary.

I added few comments, mostly nit picks.

File: model/build-processor/src/main/java/org/keycloak/models/map/processor/GenerateEntityImplementationsProcessor.java [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+                pw.println("    public enum Field {");
```
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/build-processor/src/main/java/org/keycloak/models/map/processor/GenerateEntityImplementationsProcessor.java:163]: No, this is expected to be leveraged in the configurations of the tree storage later on.

File: model/build-processor/src/main/java/org/keycloak/models/map/processor/GenerateEntityImplementationsProcessor.java (hunk 1 above)
Comment (hmlnarik) [reply model/build-processor/src/main/java/org/keycloak/models/map/processor/GenerateEntityImplementationsProcessor.java:162]: good idea

File: model/map/src/main/java/org/keycloak/models/map/common/Serialization.java (hunk 2 above)
Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/common/Serialization.java:55]: To ensure that the JSONs contain proper typing of values in case of the `EnumMap<Field, Object>` which is necessary for deserialization. It adds type information into the serialized form of the `Object` values. When omitted, the values are deserialized into default collections, e.g. into a singleton read-only set rather than a proper `ArrayList`.

File: model/build-processor/pom.xml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>keycloak-model-build-processor</artifactId>
+    <name>Keycloak Model Java Annotations and Processor</name>
+    <description/>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.keycloak</groupId>
+            <artifactId>keycloak-services</artifactId>
```
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/build-processor/pom.xml:17]: good point, it is not.

File: model/build-processor/src/main/java/org/keycloak/models/map/processor/GenerateEntityImplementationsProcessor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mhajas) [reply model/build-processor/src/main/java/org/keycloak/models/map/processor/GenerateEntityImplementationsProcessor.java:163]: In this case should we add it to separate Java file? It seems a little bit non-intuitive that it is present in one of implementations.

[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/build-processor/src/main/java/org/keycloak/models/map/processor/GenerateEntityImplementationsProcessor.java:164]: This may be reasonable, but I'd do that only later to keep this PR small

File: model/map/src/main/java/org/keycloak/models/map/common/Serialization.java (hunk 2 above)
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/common/Serialization.java:55]: I see. Thanks for the explanation.

**Review by mhajas — APPROVED** [review]
Thanks for fixes @hmlnarik. Approving.
